ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs

```diff
@@ -297,9 +297,35 @@ protected internal override void VisitNewObj(NewObj inst)
 				inst.ReplaceWith(block);
 				return;
 			}
+			if (TransformDelegateCtorLdVirtFtnToLdVirtDelegate(inst, out LdVirtDelegate ldVirtDelegate)) {
+				context.Step("new Delegate(target, ldvirtftn Method) -> ldvirtdelegate Delegate Method(target)", inst);
+				inst.ReplaceWith(ldVirtDelegate);
+				return;
+			}
 			base.VisitNewObj(inst);
 		}
 
+		/// <summary>
+		/// newobj Delegate..ctor(target, ldvirtftn TargetMethod(target))
+		/// =>
+		/// ldvirtdelegate System.Delegate TargetMethod(target)
+		/// </summary>
+		bool TransformDelegateCtorLdVirtFtnToLdVirtDelegate(NewObj inst, out LdVirtDelegate ldVirtDelegate)
+		{
+			ldVirtDelegate = null;
+			if (inst.Method.DeclaringType.Kind != TypeKind.Delegate)
+				return false;
+			if (inst.Arguments.Count != 2)
+				return false;
+			if (!(inst.Arguments[1] is LdVirtFtn ldVirtFtn))
+				return false;
+			if (!inst.Arguments[0].Match(ldVirtFtn.Argument).Success)
```

Comment:
It's wrong to transform `new Delegate(call SideEffect(), ldvirtftn TargetMethod(call SideEffect))`.
Generally, every `Match` call for deduplication needs to be combined with a `SemanticsHelper.IsPure` call.